### PR TITLE
Derive MFSDP v2 checkpoint chunk metadata from the parameter layout

### PR DESCRIPTION
**Pull request by ahmadki**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Replace the generic uneven-DTensor helper in the experimental Megatron-FSDP DCP path with an analytic, collective-free equivalent. This answers the "we may need to rewrite it from first principles for just what v2 needs" comment on #6024.

The generic helper recovers each shard's offset with one `all_gather_object` per DTensor. This path does not need to ask: `GlobalLayout` already records every parameter's global element offset, and its `DBuffer` records this rank's owned element range, so each shard's chunk follows by arithmetic.

This adds an experimental `uneven_dtensor` module that derives each parameter's chunk from its parameter group's main-weight buffer, keys it by the FQN torch's DCP state-dict helpers use, and attaches DCP's `__create_chunk_list__` / `__create_write_items__` hooks to the model and optimizer state dicts in one pass. Per-parameter optimizer state is allocated as `zeros_like` of a sharded parameter, so it shares that parameter's chunk. Tied parameters are keyed per FQN, since the state dict carries an entry for each name. The stable Megatron-FSDP helper is untouched, so v1 is unaffected.

The described chunks are unchanged, including the offsets reported for empty shards, so checkpoints stay byte-identical to the ones the generic helper produces. This is a behavior-preserving refactor that removes one collective per DTensor from both save and load.

## Dependencies

Stacked on #6024, which adds the experimental DCP path this replaces the metadata step of.

## Testing

New `tests/unit_tests/distributed/mfsdp_v2/test_uneven_dtensor.py`: the analytic metadata equals what the gather-based helper computes, computing it issues no collective (the collectives are monkeypatched to raise), and every parameter's chunks tile its global shape exactly once. Runs on 1 and 2 ranks.

Byte-identity was also checked out-of-tree on 2× RTX 6000 Ada by saving the same trained model and optimizer through both paths and comparing the per-rank `.distcp` payloads by sha256 plus the `.metadata` `state_dict_metadata` / `planner_data` / `storage_data`, for fp32 and for bf16 with fp32 master weights: identical. A DP=2 → DP=1 reshard load was verified bit-exact.

## Issue tracking

Linked issue: Related to #5534

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests 
- [x] I have added proper typing to my code
- [x] I have added relevant documentation
- [x] I have run the autoformatter.sh on my PR

### CODE REVIEW

**Review by wujingyue — COMMENTED** [review]
Same comment as https://github.com/NVIDIA/Megatron-LM/pull/6215#pullrequestreview-4846940404

**Review by wujingyue — COMMENTED** [review]
Non-blocking question: did you consider and give up on https://github.com/pytorch/pytorch/pull/189492?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/uneven_dtensor.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+
+def _chunk(row_offset: int, rows: int, shape: torch.Size) -> ChunkStorageMetadata:
+    """Build chunk metadata for a dim-0 row range of a tensor of shape ``shape``."""
+    # Flat placements only shard dim 0, so every other dimension is fully owned from offset 0.
+    return ChunkStorageMetadata(
+        offsets=(row_offset, *(0,) * (len(shape) - 1)), sizes=(rows, *shape[1:])
+    )
+
+
+def _attach_to_parameter_state(
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/uneven_dtensor.py]: Naming: state may be an optimizer state as well?

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/uneven_dtensor.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    state: dict[str, Any], metadata_by_fqn: dict[str, ChunkStorageMetadata]
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/uneven_dtensor.py]:
```suggestion
    state_dict: dict[str, Any], metadata_by_fqn: dict[str, ChunkStorageMetadata]
```
IIUC, the first parameter is a dictionary of states, not a state.

File: tests/unit_tests/distributed/mfsdp_v2/test_uneven_dtensor.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+    device = distributed_setup.device
+    mesh = init_device_mesh(device.type, (distributed_setup.world_size,))
+    model = _TiedModel().to(device=device)
+    # A tied weight cannot belong to two FsdpModules, so the parent owns the whole group.
+    with fully_shard_context(device=device):
+        fully_shard(model, mesh=mesh, placements=_placements())
+
+    metadata_by_fqn = chunk_metadata_by_fqn(model)
+
+    assert {"fc1.weight", "fc2.weight"} <= metadata_by_fqn.keys()
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_uneven_dtensor.py]: TIL: I didn't know <= can be applied on sets!

File: tests/unit_tests/distributed/mfsdp_v2/test_uneven_dtensor.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    without a chunk, which makes the save raise for that entry.
+    """
+    device = distributed_setup.device
+    mesh = init_device_mesh(device.type, (distributed_setup.world_size,))
+    model = _TiedModel().to(device=device)
+    # A tied weight cannot belong to two FsdpModules, so the parent owns the whole group.
+    with fully_shard_context(device=device):
+        fully_shard(model, mesh=mesh, placements=_placements())
+
+    metadata_by_fqn = chunk_metadata_by_fqn(model)
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_uneven_dtensor.py]: https://testing.googleblog.com/2013/08/testing-on-toilet-test-behavior-not.html

Could we move these tests to test_checkpoint.py and check the observable behavior instead? For example, the saved checkpoint contains these keys and/or certain states have the right metadata.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/uneven_dtensor.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+    """
+    # Parameters key this map by identity because tensor equality is elementwise. The map is
+    # consumed below while ``model`` still holds every parameter, so the ids stay valid.
+    metadata_by_parameter = {
+        id(fsdp_parameter.sharded): _dbuffer_chunk_metadata(parameter_group.main_weight, index)
+        for module in model.modules()
+        if isinstance(module, FsdpModule)
+        for parameter_group in module.parameter_groups
+        for index, fsdp_parameter in enumerate(parameter_group.fsdp_parameters)
+    }
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/uneven_dtensor.py]: https://google.github.io/styleguide/pyguide.html#27-comprehensions--generator-expressions

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/uneven_dtensor.py (hunk 1 above)
Comment (ahmadki) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/uneven_dtensor.py]: Renamed to `_attach_by_fqn`

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/uneven_dtensor.py (hunk 2 above)
Comment (ahmadki) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/uneven_dtensor.py]: Fixed

File: tests/unit_tests/distributed/mfsdp_v2/test_uneven_dtensor.py (hunk 4 above)
Comment (ahmadki) [reply tests/unit_tests/distributed/mfsdp_v2/test_uneven_dtensor.py]: I removed `test_uneven_dtensor.py` and moved everything to `test_checkpoint.py`. Three of the tests now go through `save_checkpoint` and assert on what the checkpoint records.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/uneven_dtensor.py (hunk 5 above)
Comment (ahmadki) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/uneven_dtensor.py]: Fair, rewritten as an explicit loop

**Comment by ahmadki** [comment]
> Non-blocking question: did you consider and give up on [pytorch/pytorch#189492](https://github.com/pytorch/pytorch/pull/189492)?

I wasn't aware of the PR. This is exactly what we need, `CheckpointableTensor` would let Megatron-FSDP describe a packed shard without touching DCP's `__create_chunk_list__` / `__create_write_items__` .

But I'm not sure about the status of the PR. it was closed without merging !

Adopting it later is a change to one module: `chunk_metadata_by_fqn` already produces the offsets and sizes, and only `_attach_dcp_hooks` would be replaced. I'm happy to revisit if the PR is revived.

**Comment by wujingyue** [comment]
> But I'm not sure about the status of the PR. it was closed without merging !

> @[pytorchmergebot](https://github.com/pytorchmergebot) pytorchmergebot added [Merged](https://github.com/pytorch/pytorch/issues?q=state%3Aopen%20label%3AMerged) and removed [merging](https://github.com/pytorch/pytorch/issues?q=state%3Aopen%20label%3Amerging) labels [on Jul 16](https://github.com/pytorch/pytorch/pull/189492#event-28093904738)

I believe it's merged: https://github.com/pytorch/pytorch/blob/8f988c9c6b3586efbc00a981d9d8cac11f26bcdb/torch/distributed/checkpoint/protocol.py#L14. 

The torch repo doesn't use the native GitHub flow to merge so all merged PRs are shown "closed with unmerged commits". Yes, that's confusing. 

cc @ahmadki 

I'll review this PR anyway.

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/uneven_dtensor.py [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+    shape = buffer.layout.tensor_shapes[index]
+    row_size = non_leading_numel(shape)
+    owned_range = buffer._get_owned_range(index)
+    if owned_range is None:
+        # Flat placements hand ranks contiguous buffer ranges in rank order, so the rows this rank
+        # does not own but that precede its range belong to lower ranks. Reporting that count keeps
+        # an empty chunk's offset equal to what a gather of all shard sizes would produce; DCP
+        # ignores it either way, because a zero-size chunk neither writes nor reads anything.
+        preceding_numel = buffer.offset - buffer.layout.tensor_to_offset[index]
+        preceding_numel = min(max(preceding_numel, 0), shape.numel())
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/uneven_dtensor.py:76]: When owend_range is None, preceding_numel is always 0 or `shape.numel()`? If so, I feel code can be simplified to something like:

```
if buffer.offset < tensor_to_offset[index]:
  0
else:
  shape.numel()
```

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/uneven_dtensor.py [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+            :func:`~torch.distributed.checkpoint.state_dict.get_model_state_dict`.
+        optimizer_state_dict: Optimizer state dict from
+            :func:`~torch.distributed.checkpoint.state_dict.get_optimizer_state_dict`.
+    """
+    metadata_by_fqn = chunk_metadata_by_fqn(model)
+    _attach_by_fqn(model_state_dict, metadata_by_fqn)
+    _attach_by_fqn(optimizer_state_dict[_OPTIMIZER_STATE_KEY], metadata_by_fqn)
+
+
+def _chunk(row_offset: int, rows: int, shape: torch.Size) -> ChunkStorageMetadata:
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/uneven_dtensor.py:140]:
```suggestion
def _build_chunk_metadata(row_offset: int, rows: int, shape: torch.Size) -> ChunkStorageMetadata:
```

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/uneven_dtensor.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+    """
+    # Parameters key this map by identity because tensor equality is elementwise. The map is
+    # consumed below while ``model`` still holds every parameter, so the ids stay valid.
+    metadata_by_parameter: dict[int, ChunkStorageMetadata] = {}
+    for module in model.modules():
+        if not isinstance(module, FsdpModule):
+            continue
+        for parameter_group in module.parameter_groups:
+            for index, fsdp_parameter in enumerate(parameter_group.fsdp_parameters):
+                metadata_by_parameter[id(fsdp_parameter.sharded)] = _dbuffer_chunk_metadata(
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/uneven_dtensor.py:108]:
```suggestion
                metadata_by_parameter[fsdp_parameter.sharded] = _dbuffer_chunk_metadata(
```
What does `id` buy us?

**Review by wujingyue — COMMENTED** [review]
Non-test LGTM. Reviewing the tests now...

File: tests/unit_tests/distributed/mfsdp_v2/test_checkpoint.py [hunk 9]
```diff
@@ -58,20 +121,83 @@ def _build_sharded(
     return model, optimizer
 
 
+def _build_packed_sharded(
+    mesh: DeviceMesh, device: torch.device, *, zero_init: bool = False
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_checkpoint.py:125]: Apparently, it's never called with zero_init=True.

File: tests/unit_tests/distributed/mfsdp_v2/test_checkpoint.py [hunk 10]
```diff
@@ -200,3 +326,166 @@ def test_checkpoint_roundtrip_flat_dp(
...
+    model_snapshot, optimizer_snapshot = _snapshot_state(model, optimizer)
+    assert {"fc1.weight", "fc2.weight"} <= model_snapshot.keys(), "The tie should surface twice"
+
+    with TempNamedDir(tmp_path_dist_ckpt / "tied", sync=True) as checkpoint_dir:
+        save_checkpoint(model, optimizer, checkpoint_dir)
+        saved_chunks = _saved_chunks(checkpoint_dir)
+        assert {"model.fc1.weight", "model.fc2.weight"} <= saved_chunks.keys()
+        assert saved_chunks["model.fc1.weight"] == saved_chunks["model.fc2.weight"]
+
+        model, optimizer = _build_tied_sharded(mesh, device, zero_init=True)
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_checkpoint.py:425]: Why zero_init=True?

File: tests/unit_tests/distributed/mfsdp_v2/test_checkpoint.py [hunk 11]
```diff
@@ -20,8 +24,20 @@
...
+)
+from megatron.core.distributed.fsdp.src.megatron_fsdp.uneven_dtensor import (
+    gather_and_compute_chunk_metadata,
+    preprocess_state_dict_for_uneven_dtensor,
+)
 from tests.unit_tests.dist_checkpointing import TempNamedDir
 
+# Collectives the chunk metadata must not need. Megatron-FSDP's stable helper recovers each
+# shard's offset with one all_gather_object per DTensor; this path derives it from the layout.
+_COLLECTIVE_NAMES = ("all_gather", "all_gather_object", "all_gather_into_tensor", "all_reduce")
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_checkpoint.py:39]: To save some scrolling, may consider move this to the only test that uses it.

File: tests/unit_tests/distributed/mfsdp_v2/test_checkpoint.py [hunk 12]
```diff
@@ -200,3 +326,166 @@ def test_checkpoint_roundtrip_flat_dp(
...
+        model, optimizer, device, param_dtype=torch.float32, in_features=4, out_features=16
+    )
+    model_state_dict = get_model_state_dict(model)
+    optimizer_state_dict = get_optimizer_state_dict(model, optimizer)
+
+    def _fail(*args, **kwargs):
+        raise AssertionError("Attaching chunk metadata must not issue a collective.")
+
+    for name in _COLLECTIVE_NAMES:
+        monkeypatch.setattr(dist, name, _fail)
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_checkpoint.py:455]: That's smart!

File: tests/unit_tests/distributed/mfsdp_v2/test_checkpoint.py [hunk 13]
```diff
@@ -200,3 +326,166 @@ def test_checkpoint_roundtrip_flat_dp(
...
+    for fqn, parameter in model.named_parameters():
+        chunk = metadata_by_fqn[fqn]
+        expected = gather_and_compute_chunk_metadata(parameter)
+        assert tuple(chunk.offsets) == tuple(expected.offsets), f"{fqn} chunk offsets"
+        assert tuple(chunk.sizes) == tuple(expected.sizes), f"{fqn} chunk sizes"
+        if chunk.sizes[0] == 0:
+            empty_offsets.append(chunk.offsets[0])
+
+    # An empty chunk at offset 0 says nothing: that is what a canonical split would report too.
+    # Only the packing gap makes a rank skip rows it does not own, so require that case.
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_checkpoint.py:487]: Can you rephrase this? I failed to parse it.

**Review by wujingyue — COMMENTED** [review]
LGTM otherwise

File: tests/unit_tests/distributed/mfsdp_v2/test_checkpoint.py [hunk 14]
```diff
@@ -200,3 +326,166 @@ def test_checkpoint_roundtrip_flat_dp(
...
+
+    for name in _COLLECTIVE_NAMES:
+        monkeypatch.setattr(dist, name, _fail)
+    attach_uneven_dtensor_metadata(model, model_state_dict, optimizer_state_dict)
+
+    assert model_state_dict.keys() == dict(model.named_parameters()).keys()
+
+
+def test_empty_shard_offsets_match_the_stable_path(distributed_setup) -> None:
+    """An empty shard is reported at the same offset the stable helper would report.
```
Comment (wujingyue) [inline tests/unit_tests/distributed/mfsdp_v2/test_checkpoint.py:462]: This falls a bit into https://testing.googleblog.com/2013/08/testing-on-toilet-test-behavior-not.html. 

What does it break if ranks don't agree on the same offset?

**Review by wujingyue — APPROVED** [review]
https://google.github.io/eng-practices/review/reviewer/speed.html#lgtm-with-comments
